[BUG] test_map_scalars_supported_key_types fails with DATAGEN_SEED=1698940723

Local repro:
```
SPARK_RAPIDS_TEST_DATAGEN_SEED=1698940723 ./run_pyspark_from_build.sh -k test_map_scalars_supported_key_types\ and\ Map\ and\ Decimal\ and\ not_null
```

```
[2023-11-02T17:07:56.061Z] ----------------------------- Captured stdout call -----------------------------
[2023-11-02T17:07:56.061Z] ### CPU RUN ###
[2023-11-02T17:07:56.061Z] ^[[31m^[[1m_ test_map_scalars_supported_key_types[Map(not_null)(Decimal(not_null)(20,2),Decimal(not_null)(20,2))] _^[[0m
[2023-11-02T17:07:56.061Z] [gw4] linux -- Python 3.8.10 /usr/bin/python
[2023-11-02T17:07:56.061Z]
[2023-11-02T17:07:56.061Z] data_gen = Map(not_null)(Decimal(not_null)(20,2),Decimal(not_null)(20,2))
[2023-11-02T17:07:56.061Z]
[2023-11-02T17:07:56.061Z]     @allow_non_gpu('WindowLocalExec')
[2023-11-02T17:07:56.061Z]     @pytest.mark.parametrize('data_gen', supported_key_map_gens, ids=idfn)
[2023-11-02T17:07:56.061Z]     def test_map_scalars_supported_key_types(data_gen):
[2023-11-02T17:07:56.061Z]         key_gen = data_gen._key_gen
[2023-11-02T17:07:56.061Z]         def query_map_scalar(spark):
[2023-11-02T17:07:56.061Z]             key_df = gen_df(spark, [("key", key_gen)], length=100).orderBy(col("key"))\
[2023-11-02T17:07:56.061Z]                 .select(col("key"),
[2023-11-02T17:07:56.061Z]                         row_number().over(Window().orderBy(col('key')))\
[2023-11-02T17:07:56.061Z]                             .alias("row_num"))
[2023-11-02T17:07:56.061Z]             key_df.select(col("key").alias("key_at_ix"))\
[2023-11-02T17:07:56.061Z]                 .where(col("row_num") == 5)\
[2023-11-02T17:07:56.061Z]                 .repartition(100)\
[2023-11-02T17:07:56.061Z]                 .createOrReplaceTempView("single_key_tbl")
[2023-11-02T17:07:56.061Z]             key_df.select(col("key").alias("key_at_ix_next")) \
[2023-11-02T17:07:56.061Z]                 .where(col("row_num") == 6) \
[2023-11-02T17:07:56.061Z]                 .repartition(100) \
[2023-11-02T17:07:56.061Z]                 .createOrReplaceTempView("single_key_tbl_next")
[2023-11-02T17:07:56.061Z]             # There will be a single row in single_key_tbl (the row that matched key_ix).
[2023-11-02T17:07:56.061Z]             # We repartition this table to create several empty tables to test empty partitions,
[2023-11-02T17:07:56.061Z]             # and also because the window operation put everything into a one partition prior.
[2023-11-02T17:07:56.061Z]             # Because this is a single key, first(ignore_nulls = true) will be deterministic.
[2023-11-02T17:07:56.061Z]             return spark.sql(
[2023-11-02T17:07:56.061Z]                 "select key_at_ix, " +
[2023-11-02T17:07:56.061Z]                 "       (select first(map(key_at_ix, 'value'), true) " +
[2023-11-02T17:07:56.061Z]                 "        from single_key_tbl)[key_at_ix], " +
[2023-11-02T17:07:56.061Z]                 # this one is on purpose using `key_at_ix` to guarantee we won't match the key
[2023-11-02T17:07:56.061Z]                 "       (select first(map(key_at_ix_next, 'value'), true) " +
[2023-11-02T17:07:56.061Z]                 "        from single_key_tbl_next)[key_at_ix] " +
[2023-11-02T17:07:56.061Z]                 "from single_key_tbl")
[2023-11-02T17:07:56.061Z] >       assert_cpu_and_gpu_are_equal_collect_with_capture(
[2023-11-02T17:07:56.061Z]             query_map_scalar,
[2023-11-02T17:07:56.061Z]             # check that GpuGetMapValue wasn't optimized out
[2023-11-02T17:07:56.061Z]             exist_classes="GpuGetMapValue",
[2023-11-02T17:07:56.061Z]             conf = {"spark.rapids.sql.explain": "NONE"})
[2023-11-02T17:07:56.061Z]
[2023-11-02T17:07:56.061Z] ^[[1m^[[31m../../src/main/python/map_test.py^[[0m:217:
[2023-11-02T17:07:56.061Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2023-11-02T17:07:56.061Z] ^[[1m^[[31m../../src/main/python/asserts.py^[[0m:402: in assert_cpu_and_gpu_are_equal_collect_with_capture
[2023-11-02T17:07:56.061Z]     assert_equal(from_cpu, from_gpu)
[2023-11-02T17:07:56.061Z] ^[[1m^[[31m../../src/main/python/asserts.py^[[0m:107: in assert_equal
[2023-11-02T17:07:56.061Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[2023-11-02T17:07:56.061Z] ^[[1m^[[31m../../src/main/python/asserts.py^[[0m:43: in _assert_equal
[2023-11-02T17:07:56.061Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2023-11-02T17:07:56.061Z] ^[[1m^[[31m../../src/main/python/asserts.py^[[0m:36: in _assert_equal
[2023-11-02T17:07:56.061Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2023-11-02T17:07:56.061Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
[2023-11-02T17:07:56.061Z]
[2023-11-02T17:07:56.061Z] cpu = Decimal('-945522642860370770.47'), gpu = Decimal('-924221809271194973.57')
[2023-11-02T17:07:56.061Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f27dfa5ca60>
[2023-11-02T17:07:56.061Z] path = [0, 'key_at_ix']
[2023-11-02T17:07:56.061Z]
[2023-11-02T17:07:56.061Z]     def _assert_equal(cpu, gpu, float_check, path):
[2023-11-02T17:07:56.061Z]         t = type(cpu)
[2023-11-02T17:07:56.061Z]         if (t is Row):
[2023-11-02T17:07:56.061Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2023-11-02T17:07:56.061Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
[2023-11-02T17:07:56.061Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
[2023-11-02T17:07:56.061Z]                 for field in cpu.__fields__:
[2023-11-02T17:07:56.061Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2023-11-02T17:07:56.061Z]             else:
[2023-11-02T17:07:56.061Z]                 for index in range(len(cpu)):
[2023-11-02T17:07:56.061Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2023-11-02T17:07:56.061Z]         elif (t is list):
[2023-11-02T17:07:56.061Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2023-11-02T17:07:56.061Z]             for index in range(len(cpu)):
[2023-11-02T17:07:56.061Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2023-11-02T17:07:56.061Z]         elif (t is tuple):
[2023-11-02T17:07:56.061Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2023-11-02T17:07:56.061Z]             for index in range(len(cpu)):
[2023-11-02T17:07:56.061Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2023-11-02T17:07:56.061Z]         elif (t is pytypes.GeneratorType):
[2023-11-02T17:07:56.061Z]             index = 0
[2023-11-02T17:07:56.061Z]             # generator has no zip :( so we have to do this the hard way
[2023-11-02T17:07:56.061Z]             done = False
[2023-11-02T17:07:56.061Z]             while not done:
[2023-11-02T17:07:56.061Z]                 sub_cpu = None
[2023-11-02T17:07:56.061Z]                 sub_gpu = None
[2023-11-02T17:07:56.062Z]                 try:
[2023-11-02T17:07:56.062Z]                     sub_cpu = next(cpu)
[2023-11-02T17:07:56.062Z]                 except StopIteration:
[2023-11-02T17:07:56.062Z]                     done = True
[2023-11-02T17:07:56.062Z]
[2023-11-02T17:07:56.062Z]                 try:
[2023-11-02T17:07:56.062Z]                     sub_gpu = next(gpu)
[2023-11-02T17:07:56.062Z]                 except StopIteration:
[2023-11-02T17:07:56.062Z]                     done = True
[2023-11-02T17:07:56.062Z]
[2023-11-02T17:07:56.062Z]                 if done:
[2023-11-02T17:07:56.062Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
[2023-11-02T17:07:56.062Z]                 else:
[2023-11-02T17:07:56.062Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
[2023-11-02T17:07:56.062Z]
[2023-11-02T17:07:56.062Z]                 index = index + 1
[2023-11-02T17:07:56.062Z]         elif (t is dict):
[2023-11-02T17:07:56.062Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
[2023-11-02T17:07:56.062Z]             # so sort the items to do our best with ignoring the order of dicts
[2023-11-02T17:07:56.062Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)
[2023-11-02T17:07:56.062Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)
[2023-11-02T17:07:56.062Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
[2023-11-02T17:07:56.062Z]         elif (t is int):
[2023-11-02T17:07:56.062Z]             assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]         elif (t is float):
[2023-11-02T17:07:56.062Z]             if (math.isnan(cpu)):
[2023-11-02T17:07:56.062Z]                 assert math.isnan(gpu), "GPU and CPU float values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]             else:
[2023-11-02T17:07:56.062Z]                 assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
[2023-11-02T17:07:56.062Z]         elif isinstance(cpu, str):
[2023-11-02T17:07:56.062Z]             assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]         elif isinstance(cpu, datetime):
[2023-11-02T17:07:56.062Z]             assert cpu == gpu, "GPU and CPU timestamp values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]         elif isinstance(cpu, date):
[2023-11-02T17:07:56.062Z]             assert cpu == gpu, "GPU and CPU date values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]         elif isinstance(cpu, bool):
[2023-11-02T17:07:56.062Z]             assert cpu == gpu, "GPU and CPU boolean values are different at {}".format(path)
[2023-11-02T17:07:56.062Z]         elif isinstance(cpu, Decimal):
[2023-11-02T17:07:56.062Z] >           assert cpu == gpu, "GPU and CPU decimal values are different at {}".format(path)
[2023-11-02T17:07:56.062Z] ^[[1m^[[31mE           AssertionError: GPU and CPU decimal values are different at [0, 'key_at_ix']^[[0m
[2023-11-02T17:07:56.062Z]
[2023-11-02T17:07:56.062Z] ^[[1m^[[31m../../src/main/python/asserts.py^[[0m:93: AssertionError
[2023-11-02T17:07:56.062Z] ----------------------------- Captured stdout call -----------------------------
[2023-11-02T17:07:56.062Z] ### CPU RUN ###
[2023-11-02T17:07:56.062Z] ### GPU RUN ###
[2023-11-02T17:07:56.062Z] ### COLLECT: GPU TOOK 0.5741064548492432 CPU TOOK 0.5611119270324707 ###
[2023-11-02T17:07:56.062Z] --- CPU OUTPUT
[2023-11-02T17:07:56.062Z] +++ GPU OUTPUT
[2023-11-02T17:07:56.062Z] @@ -1 +1 @@
[2023-11-02T17:07:56.062Z] -Row(key_at_ix=Decimal('-945522642860370770.47'), scalarsubquery()[key_at_ix]='value', scalarsubquery()[key_at_ix]=None)
[2023-11-02T17:07:56.062Z] +Row(key_at_ix=Decimal('-924221809271194973.57'), scalarsubquery()[key_at_ix]='value', scalarsubquery()[key_at_ix]=None)
```

--- a/integration_tests/src/main/python/conftest.py
+++ b/integration_tests/src/main/python/conftest.py
@@ -154,9 +154,14 @@ def pytest_runtest_setup(item):
     _inject_oom = item.get_closest_marker('inject_oom')
     datagen_overrides = item.get_closest_marker('datagen_overrides')
     if datagen_overrides:
+        try:
+            seed = datagen_overrides.kwargs["seed"]
+        except KeyError:
+            raise Exception("datagen_overrides requires an override seed value")
+
         override_seed = datagen_overrides.kwargs.get('condition', True)
         if override_seed:
-            _test_datagen_random_seed = datagen_overrides.kwargs.get('seed', _test_datagen_random_seed)
+            _test_datagen_random_seed = seed
 
     order = item.get_closest_marker('ignore_order')
     if order:

--- a/integration_tests/src/main/python/conftest.py
+++ b/integration_tests/src/main/python/conftest.py
@@ -154,7 +154,9 @@ def pytest_runtest_setup(item):
     _inject_oom = item.get_closest_marker('inject_oom')
     datagen_overrides = item.get_closest_marker('datagen_overrides')
     if datagen_overrides:
-        _test_datagen_random_seed = datagen_overrides.kwargs.get('seed', _test_datagen_random_seed)
+        override_seed = datagen_overrides.kwargs.get('condition', True)
+        if override_seed:
+            _test_datagen_random_seed = datagen_overrides.kwargs.get('seed', _test_datagen_random_seed)
 
     order = item.get_closest_marker('ignore_order')
     if order:

--- a/integration_tests/src/main/python/map_test.py
+++ b/integration_tests/src/main/python/map_test.py
@@ -192,9 +192,9 @@ def test_basic_scalar_map_get_map_value(key_gen):
 
 
 @allow_non_gpu('WindowLocalExec')
-@datagen_overrides(condition=is_before_spark_314()
+@datagen_overrides(seed=0, condition=is_before_spark_314()
                              or (not is_before_spark_320() and is_before_spark_323())
-                             or (not is_before_spark_330() and is_before_spark_331()), seed=0, reason="https://issues.apache.org/jira/browse/SPARK-40089")
+                             or (not is_before_spark_330() and is_before_spark_331()), reason="https://issues.apache.org/jira/browse/SPARK-40089")
 @pytest.mark.parametrize('data_gen', supported_key_map_gens, ids=idfn)
 @pytest.mark.xfail(condition = is_not_utc(), reason = 'xfail non-UTC time zone tests because of https://github.com/NVIDIA/spark-rapids/issues/9653')
 def test_map_scalars_supported_key_types(data_gen):

--- a/integration_tests/src/main/python/map_test.py
+++ b/integration_tests/src/main/python/map_test.py
@@ -192,7 +192,7 @@ def test_basic_scalar_map_get_map_value(key_gen):
 
 
 @allow_non_gpu('WindowLocalExec')
-@datagen_overrides(seed=0, reason='https://github.com/NVIDIA/spark-rapids/issues/9683')
+@datagen_overrides(condition=is_before_spark_340(), seed=0, reason="https://issues.apache.org/jira/browse/SPARK-40089")
 @pytest.mark.parametrize('data_gen', supported_key_map_gens, ids=idfn)
 @pytest.mark.xfail(condition = is_not_utc(), reason = 'xfail non-UTC time zone tests because of https://github.com/NVIDIA/spark-rapids/issues/9653')
 def test_map_scalars_supported_key_types(data_gen):

--- a/integration_tests/src/main/python/map_test.py
+++ b/integration_tests/src/main/python/map_test.py
@@ -192,7 +192,9 @@ def test_basic_scalar_map_get_map_value(key_gen):
 
 
 @allow_non_gpu('WindowLocalExec')
-@datagen_overrides(condition=is_before_spark_340(), seed=0, reason="https://issues.apache.org/jira/browse/SPARK-40089")
+@datagen_overrides(condition=is_before_spark_314()
+                             or (not is_before_spark_320() and is_before_spark_323())
+                             or (not is_before_spark_330() and is_before_spark_331()), seed=0, reason="https://issues.apache.org/jira/browse/SPARK-40089")
 @pytest.mark.parametrize('data_gen', supported_key_map_gens, ids=idfn)
 @pytest.mark.xfail(condition = is_not_utc(), reason = 'xfail non-UTC time zone tests because of https://github.com/NVIDIA/spark-rapids/issues/9653')
 def test_map_scalars_supported_key_types(data_gen):

--- a/integration_tests/src/main/python/spark_session.py
+++ b/integration_tests/src/main/python/spark_session.py
@@ -128,11 +128,17 @@ def is_before_spark_312():
 def is_before_spark_313():
     return spark_version() < "3.1.3"
 
+def is_before_spark_314():
+    return spark_version() < "3.1.4"
+
 def is_before_spark_320():
     return spark_version() < "3.2.0"
 
 def is_before_spark_322():
     return spark_version() < "3.2.2"
+
+def is_before_spark_323():
+    return spark_version() < "3.2.3"
 
 def is_before_spark_330():
     return spark_version() < "3.3.0"

